test/e2e/sync_lb_fail_test.go

```diff
@@ -0,0 +1,125 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var icName = types.NamespacedName{Namespace: "openshift-ingress", Name: "synclb"}
+
+var badLBService = corev1.Service{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "router-" + icName.Name,
+		Namespace: "openshift-ingress",
+		Annotations: map[string]string{
+			// this annotation is intentionally set to an invalid value, which
+			// should cause a SyncLoadBalancerFailed event
+			"service.beta.kubernetes.io/aws-load-balancer-proxy-protocol": "nonsense",
+		},
+		Labels: map[string]string{
+			"app": "router",
+			"ingresscontroller.operator.openshift.io/owning-ingresscontroller": icName.Name,
+			"router": "router-" + icName.Name,
+		},
+	},
+	Spec: corev1.ServiceSpec{
+		Type: corev1.ServiceTypeLoadBalancer,
+		Selector: map[string]string{
+			"ingresscontroller.operator.openshift.io/deployment-ingresscontroller": icName.Name,
+		},
+		Ports: []corev1.ServicePort{
+			{
+				Name:       "http",
+				Port:       80,
+				Protocol:   corev1.ProtocolTCP,
+				TargetPort: intstr.IntOrString{IntVal: 80},
+			},
+			{
+				Name:       "https",
+				Port:       443,
+				Protocol:   corev1.ProtocolTCP,
+				TargetPort: intstr.IntOrString{IntVal: 443},
+			},
+		},
+	},
+}
+
+func TestSyncLBFailEvent(t *testing.T) {
+	if infraConfig.Status.Platform != configv1.AWSPlatformType {
+		t.Skip("test skipped on non-aws platforms")
+		return
+	}
+	if err := kclient.Create(context.TODO(), &badLBService); err != nil {
+		t.Fatalf("failed to create service: %v", err)
+	}
+	defer kclient.Delete(context.TODO(), &badLBService)
+	ic := newLoadBalancerController(icName, icName.Name+dnsConfig.Spec.BaseDomain)
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+	defer kclient.Delete(context.TODO(), ic)
+	eventList := &corev1.EventList{}
+	err := wait.PollImmediate(1*time.Second, 5*time.Minute, func() (bool, error) {
+		if err := kclient.List(context.TODO(), eventList, client.InNamespace(badLBService.ObjectMeta.Namespace)); err != nil {
+			return false, nil
+		}
+		for _, event := range eventList.Items {
+			if event.Reason == "SyncLoadBalancerFailed" {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("Timed out waiting for SyncLoadBalancerFailed event")
+	}
+	expectedConditions := []operatorv1.OperatorCondition{
```

Comment:
Shouldn't there be an expected condition for `SyncLoadBalancerFailed` in this test?

Reply:
I wanted to use `waitForIngressControllerCondition` to cut down on rewriting similar functions, but it only matches condition type and status, while matching exactly on `SyncLoadBalancerFailed` would require reading the reason on the `LoadBalancerReady` status condition. But now that I try to justify that choice, I think you're right, that I should explicitly check the reason.